### ISSUE TITLE
[SPIKE] Experiment with performance of catalog queries

### ISSUE THREAD

**Issue by dbeatty10** [dbt-labs/dbt-core#9506]
### Housekeeping

- [X] I am a maintainer of dbt-core

### Short description

For dbt-bigquery and dbt-snowflake, experiment with different settings for [`relation_count`](https://github.com/dbt-labs/dbt-core/blob/9f0bcf5666a2b2a754c60f6ece7cc04e9ec81b80/core/dbt/adapters/base/impl.py#L1174):
- how does it affect the performance of the catalog query over a range of values?
- is there a point at which the query becomes to big to be accepted for execution?
  - can we determine the size of the query to be submitted so as not to exceed the 1 MB limits set by [Snowflake](https://docs.snowflake.com/en/user-guide/query-size-limits) and [BigQuery](https://cloud.google.com/bigquery/quotas)

### Acceptance criteria

We can see a graph of the performance to run the catalog query (in seconds) on the y-axis vs. the number of selected nodes on the x-axis.

We'd generally expect it to look like one of the curves below (ideally the constant time blue one, but I'm guessing not 😉):

<img width="309" alt="image" src="https://github.com/dbt-labs/dbt-core/assets/44704949/9ec72966-299f-4d7a-ad71-ceec5abbd414">

### Impact to Adapters

Depending on the results of the experiment, we may choose to use different values for `relation_count` in dbt-bigquery and/or dbt-snowflake. Alternatively, we may choose to change our implementation in some way.

### Context

The work was initially performed in https://github.com/dbt-labs/dbt-core/issues/8521 / https://github.com/dbt-labs/dbt-core/pull/8648.

Then https://github.com/dbt-labs/dbt-core/issues/9394 expressed expectation that we'd get the benefits of https://github.com/dbt-labs/dbt-core/pull/8648 even if more than 100 nodes are selected.

**Comment by ChenyuLInx** [comment]
@dbeatty10 why only bigquery and snowflake?

**Comment by ChenyuLInx** [comment]
Another dimension to consider: how many objects are in the schema.

**Comment by dbeatty10** [comment]
@ChenyuLInx yeah, it makes good sense to do both bigquery and snowflake. And also consider the number objects within the schema. 👍
